compiler/rmi/kotlinr/src/org/jetbrains/kotlin/rmi/kotlinr/KotlinCompilerClient.kt

```diff
@@ -0,0 +1,271 @@
+/*
+ * Copyright 2010-2015 JetBrains s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jetbrains.kotlin.rmi.kotlinr
+
+import org.jetbrains.kotlin.load.kotlin.incremental.components.IncrementalCache
+import org.jetbrains.kotlin.rmi.*
+import java.io.File
+import java.io.OutputStream
+import java.io.PrintStream
+import java.rmi.ConnectException
+import java.rmi.Remote
+import java.rmi.registry.LocateRegistry
+import java.util.concurrent.Semaphore
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import kotlin.concurrent.read
+import kotlin.concurrent.thread
+import kotlin.concurrent.write
+import kotlin.platform.platformStatic
+import kotlin.reflect.KProperty1
+
+fun Process.isAlive() =
+        try {
+            this.exitValue()
+            false
+        }
+        catch (e: IllegalThreadStateException) {
+            true
+        }
+
+public class KotlinCompilerClient {
+
+    companion object {
+
+        val DAEMON_STARTUP_TIMEOUT_MS = 10000L
+
+        private fun connectToService(compilerId: CompilerId, daemonOptions: DaemonOptions, errStream: PrintStream): CompileService? {
+
+            val compilerObj = connectToDaemon(compilerId, daemonOptions, errStream) ?: return null // no registry - no daemon running
+            return compilerObj as? CompileService ?:
+                throw ClassCastException("Unable to cast compiler service, actual class received: ${compilerObj.javaClass}")
+        }
+
+        private fun connectToDaemon(compilerId: CompilerId, daemonOptions: DaemonOptions, errStream: PrintStream): Remote? {
+            try {
+                val daemon = LocateRegistry.getRegistry("localhost", daemonOptions.port)
+                        ?.lookup(COMPILER_SERVICE_RMI_NAME)
+                if (daemon != null)
```

Comment:
We prefer to use braces in `if` except one-line ifs: `if (cond) return/break/continue`